Collect syntax productions in language spec into syntax.rst file

dlongnecke-cray opened the pull request:
This PR adds a script to `doc/util` that collects all the productions listed throughout the RST language spec. The script outputs the productions in alphabetical and application (breadth-first) order into the spec file `syntax.rst`. This script restores behavior that was removed when the language spec was moved to RST in #14303.

The script checks to make sure that all listed productions are applied at least once elsewhere in the spec. Add the `import-statement` production to the RHS of the `statement` production because it was not applied elsewhere.

Rename a few bad non-terminals that did not have corresponding productions.

---

Testing:

- [x] `ALL` on `linux64` when `COMM=none`
- [x] Updated docs rendered locally
- [x] Updated docs hosted for review

dlongnecke-cray commented:
Hosted docs with new syntax page can be viewed at: http://web.us.cray.com/~dlongnecke/restore-chapel-grammar-rst/language/spec/syntax.html#productions-listed-application-order

dlongnecke-cray commented:
Hi @mppf since you removed the original script, would you mind doing a quick review to see if things look OK?

Hi @lydia-duncan since you're an expert on the docs, would you review the changes to the `doc/Makefile` and the changes to the spec files?

mppf commented on doc/Makefile:
```diff
@@ -17,6 +17,9 @@ CHPL2RSTOPTS = --output=rst --prefix=$(CHPL_DOC_PRIMERS_DIR) --link=$(WEB_DOC_DI
 CHPL2RSTOPTS_HELLO = --output=rst --prefix=$(CHPL_DOC_HELLO_DIR) --link=$(WEB_DOC_DIR)$(WEB_DOC_VERSION)
 
+SPECSRCDIR = $(SOURCEDIR)/language/spec
+GETSYNTAX = ../util/config/run-in-venv.bash ./util/collect-syntax.py
```
Something needs to change here after PR 16694 - I recommend you rebase now that PR is merged. However I don't see any reason why collect-syntax.py needs to run in the venv.

mppf commented on doc/Makefile:
```diff
@@ -42,8 +45,13 @@ man-chapel: FORCE
+collect-syntax: clean-collect-syntax
+	@echo
+	@echo "Collecting syntax from '$(SPECSRCDIR)' into '$(SPECSRCDIR)/syntax.rst'"
+	$(GETSYNTAX) rst/language/spec --outdir $(SPECSRCDIR)
```
since it is just run once, why have a GETSYNTAX variable? AFAIK this should just be ./util/collect-syntax.py

mppf commented on doc/util/collect-syntax.py:
```diff
@@ -0,0 +1,327 @@
+"""
+
+
+from __future__ import print_function
```
should not be needed now that python3 is a requirement for docs build

mppf commented on doc/util/collect-syntax.py:
```diff
@@ -0,0 +1,327 @@
+  except Exception as e:
+    traceback.print_exc()
+    err = 1
+  sys.exit(err)
```
I am susprised you need this try/except/exit. Doesn't all that happen by default?

dlongnecke-cray replied:
I wasn't sure, I just wanted to make it so you could collect the exit code, but perhaps not necessary :).

mppf commented on doc/util/collect-syntax.py:
```diff
@@ -0,0 +1,327 @@
+
+
+"""
+collect syntax
```
you can do better than that for a comment summarizing this script!

mppf commented on doc/util/collect-syntax.py:
```diff
@@ -0,0 +1,327 @@
+_rstfiles = None
+_outdir = None
+_outfile = None
+_verbose = False
```
It seems odd to me that all of these variables begin with `_`. Is there some Python convention I am not aware of?

dlongnecke-cray replied:
It's just a habit of indicating weak internal use that isn't really necessary in this script (I forced myself not to do it for functions). I normally don't use many global variables in Python, is there a convention we'd like to use for them? Or just named like anything else?

mppf replied:
I don't have any particular opinion on it.

mppf commented on doc/util/collect-syntax.py:
```diff
@@ -0,0 +1,327 @@
+  return
+
+
+def main(args):
```
You should not need to use the venv/chpldeps to run this script but we are using a convention there that `main` has no arguments (for things that are run through chpldeps). You can rename this one as `main_arguments`  and create `main` that is the body of the `if __name__ == '__main__'`. While it does not matter for this script I think it would be nice to follow the convention here.

mppf commented on doc/util/collect-syntax.py:
```diff
@@ -0,0 +1,327 @@
+_match_syntax_begin = re.compile(r"\s*.. code-block:: syntax\s*\n")
+_match_syntax_end = re.compile(r"\S+")
+_match_production = re.compile(r" {3}(?P<ident>([a-zA-Z0-9._-]+)):.*")
+_match_expansion = re.compile(r" {3} +")
```
Is this really looking for 3 spaces? That seems a bit too fragile to me. In `doc/util/extract-rst-tests.py` I added a function that knows how to read an rst block (based on indenting). You could use that here I think. It seems to me we could/should use `read_block` or something like it both for the `code-block:: syntax` and also for the block from e.g. `expression:`.

dlongnecke-cray replied:
I will look into it, but as far as I know every single syntax block is indented 3 spaces on the first level.

mppf replied:
Right, but that doesn't mean the tooling should break if that stops being true in the future.

dlongnecke-cray replied:
Only thing left for me to fix now is this.+ the global variable names (if we decide we have a problem with them).

lydia-duncan reviewed (approved):
The things you asked me to look at look good to me :)

dlongnecke-cray commented:
Hi @mppf, I believe I've addressed all your feedback, including making the parsing more flexible in the presence of varying indentation levels. The rules for the parsing are described in the comment at the top of the script.

Look good?
